Fix: Prevent the daemon from starting if the sync box is not detected

esavary opened the pull request:
- At startup, the daemon checks if the trigger box is plugged in, throws an error and stops if this is not the case.
 - Add a check in the SOPs for the status of the daemon and precise the order in which to plug the USB cables.

esavary committed:
ENH: Add error when the triggerbox is not plugged

- At startup the deamon checks if the trigger box is plugged in and throws an error and stops if this is not the case.

esavary committed:
FIX: change the order to plug USB cables

- Change the order to plug the USB cable.
- Add a check for the status of the daemon.

esavary committed:
FiX: style issues in forward-trigger-service.py

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -25,12 +25,13 @@
 
 
-LISTEN = 2023
+LISTEN = 8888
```
Why this change?

esavary replied:
I changed it back to 2023.

oesteban replied:
The rationale is that you may have a jupyter notebook running and we don't want to have any conflicts.

oesteban commented on docs/data-collection/pre-session.md:
```diff
@@ -191,6 +191,8 @@ Instructions of operations to be performed before the participant arrival, **bef
     
         ![switch_screen](../assets/images/screen_switch.jpg)
+      
+- [ ] Connect the NEUROSPEC adapter (pink USB cable) to one of the USB port of the laptop *{{ secrets.hosts.psychopy | default("███") }}*. **This cable must always be connected last.**
```
```suggestion
- [ ] Connect the MMBT-S Trigger Interface Box adapter (pink USB cable) to one of the USB ports of the laptop *{{ secrets.hosts.psychopy | default("███") }}*.
    ![neurospec](../assets/images/neurospec.jpg)
    
    !!! danger "The MMBT-S Trigger Interface MUST be connected to the laptop AFTER the trigger USB cable coming from the SyncBox. "
```

oesteban commented on docs/data-collection/pre-session.md:
```diff
@@ -63,8 +63,8 @@ Instructions of operations to be performed before the participant arrival, **bef
-    ![neurospec](../assets/images/neurospec.jpg)
-- [ ] Connect that NEUROSPEC adapter to one of the USB port of the laptop *{{ secrets.hosts.psychopy | default("███") }}*.
+    ![neurospec](../assets/images/neurospec.jpg) ** Do not connect the NEUROSPEC parallel-to USB-converter to *{{ secrets.hosts.psychopy | default("███") }}* now. It must always be connected after the USB cable of the sync box.**
+
```
```suggestion
```

esavary committed:
Update docs/data-collection/pre-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update docs/data-collection/pre-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
FIX: change port number

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -52,13 +53,10 @@ async def handle_client(
 ) -> None:
     """
     Handle client connections.
-
```
Please add these lines back, the convention of docstrings is that the first line is a one-line short description that starts with an action and ends in period. Then an empty line and then the long description may or may not expand through several lines.

BTW, I'm noticing that the parameters docstring is not numpy/napoleon compliant. I will fix that.

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -52,13 +53,10 @@ async def handle_client(
     Handle client connections.
-
     Read incoming data from the client and put it into the async queue.
-
```
Ditto

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -52,13 +53,10 @@ async def handle_client(
     Parameters:
         reader (asyncio.StreamReader): The client's stream reader.
         async_q (janus.AsyncQueue[int]): The async queue for storing incoming signals.
-
```
Ditto (and so on, so forth all the way to the end)

oesteban commented on docs/data-collection/pre-session.md:
```diff
@@ -200,6 +203,7 @@ Instructions of operations to be performed before the participant arrival, **bef
     
 - [ ] Configure the display settings of the laptop to mirror outputs and set a resolution of 800x600 for both screens.
 - [ ] Double check that the IP address corresponding to the ethernet interface of the *{{ secrets.hosts.psychopy | default("███") }}* laptop is correct. You can either run `ifconfig -a` or use the GUI. Make sure the IP/mask is **100.1.1.2/24**, and the protocol is IP version 4. Execute `ping 100.1.1.1` to see if the ET is responding to echoes.
+- [ ] Check that the service to synchronize the triggers is on with `sudo systemctl status forward-trigger`. If not, run `sudo systemctl status forward-trigger` and recheck the status.
```
```suggestion
- [ ] Check that the service to synchronize the triggers is up with `sudo systemctl status forward-trigger`.

    !!! warning "If the service is down, manually force its start"
    
        - [ ] Run `sudo systemctl status forward-trigger`
        - [ ] Recheck the status with `sudo systemctl status forward-trigger`.
    
    !!! important "These commands are executed with `sudo`"
    
        The console will prompt you for the common user password: `{{ secrets.login.password_hos68752 | default("****") }}`

```

oesteban commented on docs/data-collection/pre-session.md:
```diff
@@ -63,8 +63,6 @@ Instructions of operations to be performed before the participant arrival, **bef
     ![mac_setup](../assets/images/mac_setup.png) 
 - [ ] Connect one end of the parallel cable on the parallel plug on the back side of the STP100D unit of the BIOPAC and the other end to the parallel port of the NEUROSPEC parallel-to USB-converter.
```
Please revise this (I can't using the PR review utility because you did not modify this line):

1. Let's be more precise and replace any instance of "NEUROSPEC <blah blah>" with "MMBT-S Trigger Interface Box adapter" or "MMBT-S Trigger Interface Box adapter (N-shaped pink box)".
2. Let's be more precise and say that we want to connect the parallel cable to the 25-pin socket at the back of the SPT100D.

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
+    # Check if the USB device is connected
+    initial_usb_connection = check_usb_device_connected(
+        "07c0", "0101"
+    )  # for testing "0bc2", "2322"
```
please move these magic numbers to the top of the file, by setting some static-ish variables (python doesn't have that):

```Python
USB_MMBTS_DEVICE_ID = ("07c0", "0101")
"""Device ID of the Neurospec's MMBT-S trigger adaptor."""
USB_TESTING_DEVICE_ID = ("0bc2", "2322")
"""Device ID for testing purposes."""
```

and then
```suggestion
    ensure_usb_device_connected(*USB_MMBTS_DEVICE_ID)
```

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
 
+def check_usb_device_connected(usb_vendor_id, usb_product_id):
+    """
+    Check if a USB device with the specified vendor and product IDs is connected.
+    Parameters:
+        usb_vendor_id (str): The USB vendor ID.
+        usb_product_id (str): The USB product ID.
+    Returns:
+        bool: True if the device is connected, False otherwise.
```
```suggestion
    Check if a USB device with the specified vendor and product IDs is connected.
    
    Parameters
    ----------
    usb_vendor_id : :obj:`str`
        Vendor ID of a USB device.
    usb_product_id : :obj:`str`
        Product ID of a USB device.

    Returns
    -------
    connected : :obj:`bool`
        ``True`` when the device is connected. If the device is not connected, this
        function raises a :obj:`RuntimeError`.

```

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
     logging.info("Scanner trigger received")
 
 
+def check_usb_device_connected(usb_vendor_id, usb_product_id):
```
Since this function raises exception if not connected, perhaps it would be better to call it differently (this is just a suggestion):

```suggestion
def ensure_usb_device_connected(usb_vendor_id, usb_product_id):
```

oesteban commented on code/synchronization/forward-trigger-service.py:
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
-
     """
+    # Check if the USB device is connected
+    initial_usb_connection = check_usb_device_connected(
```
Why keeping a variable if this is expected to crash if the device is not present?

```suggestion
    ensure_usb_device_connected(
```

oesteban reviewed (commented):
this is starting to look great! thanks for this :)

esavary committed:
Update docs/data-collection/pre-session.md

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/synchronization/forward-trigger-service.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/synchronization/forward-trigger-service.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
Update code/synchronization/forward-trigger-service.py

Co-authored-by: Oscar Esteban <code@oscaresteban.es>

esavary committed:
FIX: precise the name of NEUROSPEC device

- Replace NEUROSPEC with MMBT-S Trigger Interface Box adapter.
- Precise the emplacement of the connexion of the parallel cable.
-  Fix typos.

esavary committed:
FIX: fix code style

esavary committed:
FIX: fix code style

esavary commented:
Thanks for the review! I've integrated the comments in the last commits.

oesteban committed:
sty: proper numpy docstrings, use hex for usb IDs

oesteban commented:
@esavary I'm merging because the code looks good and I'm sure you have run `mkdocs serve` and have checked that the markdown renders correctly :)
